src/signing_key.py

```diff
@@ -0,0 +1,115 @@
+#!/usr/bin/env python3
+
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Synapse signing key."""
+
+
+import logging
+import typing
+
+import ops
+
+import synapse
+from charm_state import CharmState
+
+logger = logging.getLogger(__name__)
+
+
+def signing_key_path(charm_state: CharmState) -> str:
+    """Get signing key path.
+
+    Args:
+        charm_state: charm state.
+
+    Returns:
+        signing key path as string.
+    """
+    return f"/data/{charm_state.synapse_config.server_name}.signing.key"
+
+
+def get_signing_key_secret_content(
+    peer_relation: typing.Optional[ops.Relation], charm: ops.CharmBase
+) -> typing.Optional[str]:
+    """Get signing key secret content.
+
+    Args:
+        peer_relation: synapse peer relation.
+        charm: charm instance.
+
+    Returns:
+        Content as string.
+    """
+    content = None
+    if not peer_relation:
```

Comment:
If we check for the relation in reconcile, we can keep the check here to be defensive, but since we can assume the relation is there, we therefore can throw an exception (and remove the optional for the relation)